src/main/scala/scalaz/ioqueue/Queue.scala

```diff
@@ -205,38 +205,74 @@ class Queue[A] private (
         takers.dequeueOption match {
           case None =>
             val (addToQueue, surplusValues) = capacity.fold((as, Iterable.empty[A]))(as.splitAt)
-            val (complete, putters) =
-              if (surplusValues.isEmpty)
-                p.complete(true) -> IQueue.empty
-              else
-                IO.now(false) -> IQueue.empty.enqueue(surplusValues -> p)
-
-            complete -> Surplus(IQueue.empty.enqueue(addToQueue.toList), putters)
+            if (surplusValues.isEmpty)
+              p.complete(true) -> Surplus(IQueue(addToQueue.toSeq: _*), IQueue.empty)
+            else
+              strategy match {
+                case BackPressure =>
+                  IO.now(false) -> Surplus(
+                    IQueue.empty.enqueue(addToQueue.toList),
+                    IQueue.empty.enqueue(surplusValues -> p)
+                  )
+                case Sliding =>
+                  val toQueue = (addToQueue ++ surplusValues).takeRight(surplusValues.size)
```

Comment:
```suggestion
                  val toQueue = capacity.fold(as)(as.takeRight)
```
`(addToQueue ++ surplusValues)` is just `as`.
And we can `takeRight` as much as `capacity` allows.